lib/spack/spack/build_environment.py

```diff
@@ -994,7 +1003,8 @@ def add_modifications_for_dep(dep):
             if set_package_py_globals:
                 set_module_variables_for_package(dpkg)
             # Allow dependencies to modify the module
-            dpkg.setup_dependent_package(spec.package.module, spec)
+            for module in nonshared_parents(spec.package.__class__):
+                dpkg.setup_dependent_package(module, spec)
```

Comment:
I worry that this will apply spec changes multiple times, which will be bad if they have side effects.

Reply:
Good point, although I don't know how to address that without changing the API (for all packages in builtin that define this, for example): we need to run `setup_dependent_package` on every superclass module; up to now, the `setup_dependent_package` has been free to modify the spec as it sees fit. If it only makes idempotent changes (e.g. `setattr`) then it would be fine (but we have not mandated that constraint before).